src/liballoc/borrow.rs

```diff
@@ -479,3 +479,15 @@ impl<'a> AddAssign<Cow<'a, str>> for Cow<'a, str> {
         }
     }
 }
+
+#[stable(feature = "cow_str_add_assign_char", since = "1.41.0")]
+impl AddAssign<char> for Cow<'_, str> {
+    fn add_assign(&mut self, rhs: char) {
+        if let Cow::Borrowed(lhs) = *self {
+            let mut s = String::with_capacity(lhs.len() + rhs.len_utf8());
```

Comment:
This capacity calculation seems like the wrong default. If the caller is extending a string by `+=`ing chars, it seems pretty likely that they will do more than one. It would be better to use String::reserve to take advantage of String's existing capacity heuristics.

Reply:
Can we predict a sane value to be used for reserve? @dtolnay

Reply:
Yes, the underlying RawVec type that Vec and String are based on has these amortization heuristics built in already. I think it would be worth routing this impl in some way through [RawVec::reserve](https://github.com/rust-lang/rust/blob/a44881d892fb4f4a8ed93f8f392bab942fac7a41/src/liballoc/raw_vec.rs#L519) (directly or indirectly), where used_capacity = lhs.len() and needed_extra_capacity = rhs.len_utf8().

Reply:
@dtolnay What you're suggesting is to rather use `String::reserve()` and not `String::capacity()`?

This is how it would then look like:
```rust
if let Cow::Borrowed(lhs) = *self {
	let mut s = String::new();
        s.reserve(lhs.len() + rhs.len_utf8());
	s.push(lhs);
	*self = Cow::Owned(s);
}
```

Initially I thought you wanted a preset increase and that left me puzzled because I couldn't imagine how that could be predicted. `String::reserve()`s internal implementation is already what is needed here.

Moreover, this condition: `if let Cow::Borrowed(lhs) = *self` is a setback for us to fathom further memory reservation ideas, since the allocation would happen everytime this condition is true regardless of our `String` memory allocation approach. Thoughts?

Reply:
No, I am suggesting to use RawVec::reserve. String::reserve is not sufficient because it only takes one argument. In order to amortize well, an optimal heuristic really requires two arguments, current size and requested additional capacity. The code you showed is in fact equivalent to `String::with_capacity(lhs.len() + rhs.len_utf8())` which is bad because it requires an immediate second allocation if the caller pushes a second character.

Reply:
The [current heuristic](https://github.com/rust-lang/rust/blob/e2fa9527d44522b900c739cc82faa734f877bcda/src/liballoc/raw_vec.rs#L446-L459) for RawVec::reserve is effectively `max(current_capacity*2, current_length+requested_capacity)`. So in the case of:

```rust
let mut s = String::new();
s.reserve(lhs.len() + rhs.len_utf8());
```

the `reserve` will observe current_capacity=0, current_length=0, requested_capacity=lhs+rhs, which means it allocates capacity for exactly lhs+rhs according to the heuristic.

But for this impl we want to treat lhs.len() as the current_length and rhs.len_utf8() as the requested additional capacity and get the right amortized behavior.

Reply:
@dtolnay Can we derive a utility function based off `amortized_new_size()` which would look like this:

```rust
/// Calculates the buffer's new size given that it'll hold `used_capacity +
/// needed_extra_capacity` elements. This logic is used in amortized reserve methods.
/// Returns `(new_capacity, new_alloc_size)`.
fn amortized_new_size(current_capacity: usize, used_capacity: usize, needed_extra_capacity: usize)
	-> Result<usize, TryReserveError> {

	// Nothing we can really do about these checks, sadly.
	let required_cap = used_capacity.checked_add(needed_extra_capacity)
		.ok_or(CapacityOverflow)?;
	// Cannot overflow, because `cap <= isize::MAX`, and type of `cap` is `usize`.
	let double_cap = current_capacity * 2;
	// `double_cap` guarantees exponential growth.
	Ok(cmp::max(double_cap, required_cap))
}
```?

Reply:
Or maybe the `String` type should have an inherent function like that? So that all `String`s can be handled in this manner.